Make m4-single the default floating-point precision ABI

darcagn opened the pull request:
Let's finally make `m4-single` the default. This was discussed with @QuzarDC and @ljsebald, and these were the points made:

- `m4-single` being the default in KOS/GCC is fine.
  - While it's nice that `m4-single` is a currently enabled option now, `m4-single` as a _default_ for GCC is desirable since `libgccjit`, which can be used to interface with the `rustc` compiler frontend when built with `m4-single`, does not support multilib, and is only built for the default.
  - GCC can be built with one setting as default, and KOS can use another setting as its default, though.
- More testing within KOS is needed to move to `m4-single` as default. 
  - A handful of people using it as their primary setting does not suffice for testing.
- Full compatibility in KOS with `m4-single-only` needs to remain.
- Any kos-ports incompatible with `m4-single` or `m4-single-only` is fine but this must be explicitly stated in associated documentation/readme. 
  - kos-ports checking this for you when building would be nice.
  - I've submitted a PR with such functionality: [kos-ports #90](https://github.com/KallistiOS/kos-ports/pull/90)

Some clarification around sufficient testing would be nice. As far as I know, all examples build and function just fine on `m4-single` right now. This doesn't test all code paths though. And further testing of ports would certainly be nice, since some of them aren't used very often anymore, etc.

QuzarDC reviewed (commented):
Just the one thing that I hadn't realized from our previous discussions and am a little wary about. Otherwise I think we're quite ready for this change and all the groundwork had been laid out well in advance to make this a very simply changeover.

darcagn commented on environ_dreamcast.sh:
```diff
@@ -11,9 +11,9 @@ if [[ ":$PATH:" != *":${DC_TOOLS_BASE}:"* ]]; then
-# Default the SH4 floating point precision if it isn't already set.
+# Default the SH4 floating-point precision if it isn't already set.
 if [ -z "${KOS_SH4_PRECISION}" ] ; then
-    export KOS_SH4_PRECISION="-m4-single-only"
```
Good point. I've added a test for -m4-single and a fallback to -m4-single-only if GCC doesn't support -m4-single.

QuzarDC reviewed (dismissed):
Tested the changes against a few different environments and the testing for fallback automatically triggered correctly. I couldn't find any issue building out all of kos-ports, examples, and some random bits of software. 

As @ljsebald has been very concerned about this change in the past ( see #397 ), want to make sure he's ok with it too before moving forward though. From what I can tell, in the time since then substantially more groundwork has been done to correct issues around precision including: #438 #451 #460 #518 #540 #724 .

Additionally it's been used in some of the larger port projects, a majority of our contributors seem to be using it, and DreamSDK added support some time ago (having it available in the prebuilt toolchains).

I think the important part to keep in mind on the distinction is entirely that `m4-single-only` *disallows* the use of double precision entirely. So to have relied on that would require code that used doubles and relied on the behavior that happened when they were being truncated rather than maintained at full precision.

pcercuei commented:
Let's merge this?

ljsebald commented:
I think this definitely deserves a RELNOTES mention before making this change, considering the behavior difference around double precision values.

I assume that this has been thoroughly tested everywhere (i.e, all examples, etc)?

darcagn commented:
> I think this definitely deserves a RELNOTES mention before making this change, considering the behavior difference around double precision values.

That's definitely warranted. I'll update this PR with release notes thoroughly explaining the changes.

> I assume that this has been thoroughly tested everywhere (i.e, all examples, etc)?

Yes, all examples have been tested & verified to work great using `m4-single` for quite some time. I'm not aware of any issues within KOS, its examples, kos-ports, or any external software when using `m4-single` at this time.

ljsebald commented:
> > I think this definitely deserves a RELNOTES mention before making this change, considering the behavior difference around double precision values.
> 
> That's definitely warranted. I'll update this PR with release notes thoroughly explaining the changes.
> 
> > I assume that this has been thoroughly tested everywhere (i.e, all examples, etc)?
> 
> Yes, all examples have been tested & verified to work great using `m4-single` for quite some time. I'm not aware of any issues within KOS, its examples, kos-ports, or any external software when using `m4-single` at this time.

Once the RELNOTES update is done, I'll approve then. Thanks for the confirmation.

darcagn commented:
@ljsebald @QuzarDC 

I've rebased against latest KOS changes since the last review, and added documentation of the change to `RELNOTES.md` and `CHANGELOG.md`. Please let me know if anything else needs to be addressed in the documentation!

darcagn commented:
I've updated this PR to adjust `loadable/Makefile.prefab` to respect the $KOS_SH4_PRECISION variable. Using `m4-single` here has already been successfully tested downstream in DreamShell: https://github.com/DC-SWAT/DreamShell/pull/40
